// @youi/react-native-youi/index.js
'use strict'

import { requireNativeComponent } from 'react-native'

export {
  Animated,
  AppRegistry,
  Button,
  Dimensions,
  Keyboard,
  FlatList,
  Image,
  PixelRatio,
  SafeAreaView,
  ScrollView,
  Slider,
  StyleSheet,
  Text,
  TextInput,
  TouchableHighlight,
  View,
  VirtualizedList,
  WebView
} from 'react-native'

let TextRef = requireNativeComponent('YiTextSceneNode')

import BackHandler from './js/BackHandler.js'
import ButtonRef from './js/ButtonRef.js'
import Composition from './js/Composition.js'
import DeviceInfo from './js/DeviceInfo.js'
import FairPlayDrmHandler from './js/FairPlayDrmHandler.js'
import FocusManager from './js/FocusManager.js'
import FontRegistry from './js/FontRegistry.js'
import FormFactor from './js/FormFactor.js'
import ImageRef from './js/ImageRef.js'
import Input from './js/Input.js'
import ListRef from './js/ListRef.js'
import ScrollRef from './js/ScrollRef.js'
import SliderRef from './js/SliderRef.js'
import TextInputRef from './js/TextInputRef.js'
import TimelineRef from './js/TimelineRef.js'
import Video from './js/Video.js'
import VideoRef from './js/VideoRef.js'
import ViewRef from './js/ViewRef.js'
import WebViewRef from './js/WebViewRef.js'

export {
  BackHandler,
  ButtonRef,
  Composition,
  DeviceInfo,
  FairPlayDrmHandler,
  FocusManager,
  FontRegistry,
  FormFactor,
  ImageRef,
  Input,
  ListRef,
  ScrollRef,
  SliderRef,
  TextInputRef,
  TextRef,
  TimelineRef,
  Video,
  VideoRef,
  ViewRef,
  WebViewRef
}
